commands/HTTPRetryRequest.js

```diff
@@ -0,0 +1,64 @@
+const https = require("https");
+
+let DEFAULTS = {
+  MAX_RETRIES: 5,
+  ERROR_CODES: ""
+};
+
+function HTTPRetryRequest(config, options) {
+  let _self = this;
+  _self.options = options;
+  _self.config = { ...DEFAULTS, ...config }; // overwrite defaults
+
+  if (_self.config.ERROR_CODES && _self.config.ERROR_CODES.length) {
+    // create pattern of /(5\d\d|4\d\d|9\d\d)/
+    _self.config.ERROR_CODES = _self.config.ERROR_CODES.replaceAll(",", "|"); // replace sperator , with | for regex
+    _self.config.ERROR_CODES = _self.config.ERROR_CODES.toUpperCase(); // make uppercase (x -> X)
+    _self.config.ERROR_CODES = _self.config.ERROR_CODES.replaceAll("X", "\\d"); // replace character X with \d for regex - represents a digit
+    _self.config.ERROR_CODES = new RegExp("(" + _self.config.ERROR_CODES + ")"); // wrap in () to have the alternative
+  }
+  if (!_self.config.retryCount) {
+    _self.config.retryCount = 1;
+  } else {
+    _self.config.retryCount++;
+  }
+  _self.buffer = Buffer.alloc(0);
+
+  return new Promise((resolve, reject) => {
+    let requestInstance = https.request(options, response => {
+      const innerStatusCode = response.statusCode.toString();
+      const responseInstance = response
+        .on("error", error => {
+          responseInstance.abort();
+          if (_self.config.ERROR_CODES && _self.config.ERROR_CODES.test(innerStatusCode) && _self.config.retryCount < _self.config.MAX_RETRIES) {
+            // if the status is one of the ones we want to retry, then make the same request
+            resolve(new HTTPRetryRequest(_self.config, _self.options));
+          } else {
```

Comment:
From my understanding of the issue, the error codes are not related to retries. Its different functionality.

Error codes was meant to be, if you want the task to fail on certain response codes. As right now it only fails the task if something goes wrong with making the call, and then you have to run another task after it to check the response code resultParameter.

Reply:
@tlawrie how it was described to me it was to retry only when the error codes that are inputed are found as the status code. If your understanding is different please append the task with all the details and I will do the changes.

Reply:
Hi @florentinvintila interesting ok. We may need to clarify with @morarucostel 

As far as I know we had two different tasks;

**1. [Cause the task to fail if a certain Status Code range is not achieved](https://github.com/boomerang-io/roadmap/issues/34)**
i.e. 2xx / 3xx / 4xx / 5xx then fail the task. This would most likely be a drop down of Status Code ranges. OR the flip of this. A drop down of what is considered success, i.e. the task only succeeds if a 2xx is received. This could also be a check box that states "Task only succeeds on 2xx Status Code" or something like this.

**2. [Retry the call y number of times](https://github.com/boomerang-io/worker.flow/issues/44)**
Add the ability to specify a certain number of retries, utilize if a timeout is received.

They aren't mutually related and in fact two separate outcomes. However the last time it was discussed, if both changes were done at the same time, then the testing would be less overall.

Reply:
Hi @tlawrie , you are correct, there are two mutually (at least they started as) unrelated requirements but now that I look at them, we might need to treat them as one. Treating them separately will raise some edge cases that are difficult to manage and will become more difficult for the end-user to configure them. 
Requirement number 2, from your list, is the one that is fully addressed in this PR, but now that we think of, it doesn't make sense to have retry capability when the status code is considered a `success`. 

For that reason, I think we should address both requirements together by: 
 - using a single input text field with the Status codes that are mapping the "Error" scenario. We can allow the end-user to put "4xx,5xx" or even more granular error codes "4xx,502,503". 
 - and using a second input number field with the number of retries in case the HTTP status code is coming as one of the listed values mapped as error. 
This way we keep to a minimum the number of input fields the end-user has to work with, there are no edge cases of overlapping different types of codes and we support both requirements.

Reply:
@tlawrie We had another call today with @florentinvintila and agreed on the following:
We will use three fields in order to model the solution:
 - using an input text field with the Status codes that are defined for the scenario. We can allow the end-user to put "4xx,5xx" or even more granular error codes "4xx,502,503".
 - using a drop down with the options Success or Failure, that defines the scenario selected by the end-user.
 - using a third field, input number with the number of retries in case the HTTP status code is coming as an error mapped one.

That means:
- if the end-user select a range of 2xx, `success` in the drop down and 3 in the number of retries, the task will retry at most 3 time (if it doesn't receive a 2xx http code). The task will end with failure (in the scenario where no 2xx is received within those 3 retries),
- if the end-user select a range of 4xx,5xx, `failure` in the drop down and 3 in the number of retries, the task will retry at most 3 time (if it receives a 4xx or 5xx http code). If it receives a return code in the 2xx area, it will finish and return as a successful run. Anything else (other than 4xx, 5xx and 2xx) is not retried and is closed as failed run. 
This model is covering the edge cases of custom defined api statuses by various implementations.

Reply:
Ok. That makes sense. And thanks for going through the use detail.

However, it does present a problem where this is now really complex. With three combinations of things that a user has to adjust it becomes quite tedious, particularly for the basic users. 

I am thinking we could pair it back a bit...
1. Create a separate task of Advanced HTTP Task that has all this extra logic and they share some of the same code under the covers
2. Do we really need the custom range or can we not just have a drop-down with 1xx, 2xx, 3xx, 4xx, 5xx?

Reply:
@tlawrie please take another look at it. An additional delay time was introduced between retries.

Reply:
Thanks @florentinvintila so I think we are going to have to split this out as a separate task. They can both share the same base HTTP execution but this is now super complex.

For a standard user having to come and use this task and worry about what to see and these multiple combinations of status codes and status options. 

I think we need to have an Execute `Advanced HTTP Call` because we now are taking in 4 extra options that a standard not technical no-code user is not going to want to have to deal with.

Also what happened to simplifying the error codes
> Do we really need the custom range or can we not just have a drop-down with 1xx, 2xx, 3xx, 4xx, 5xx?

Reply:
Hey @tlawrie , creating a new Advanced HTTP task will complicates a lot our uses case because we have many existing workflows relying on the current `Execute HTTP Call` and we rely heavily on the backward compatibility. If we introduce a new task, we'll have to change all our existing workflows and introduce a lot of regression. 

Instead we want to simplify the experience by:
 - setting the defaults (of these new 3 fields) in such a way that a typical "business" user doesn't even have to change or look at them (allowing backward compatibility),
 - basically the end-user doesn't need to do anything in that section and it will run just as before, so we don't introduce complexity.

Reply:
@tlawrie @morarucostel  perhaps our existing one can be renamed to Advanced HTTP Call, then we create a new basic HTTP task?

Reply:
@marcusdroy That is great idea. Just renaming the current task to Advanced HTTP Call means the existing workflows will continue to use the same task based on the id of the task. 
I believe this might work.

Reply:
Happy for a new task. I personally didn't want two tasks but at the same time, i don't think the number and logic of the options is going to be easy for a non-technical user to understand. Its quite the conundrum.

So i think if we go with the `Execute Advanced HTTP Call` and `Execute Basic HTTP Call` how do we go about ensuring that the code is reused as much as possible? does the task template for the Basic call just call the same method but doesn't pass in some of the inputs and hence they are defaulted?